data-generation/src/main/scala/thesis/LandyConsumer.scala

```diff
@@ -0,0 +1,72 @@
+package thesis
+
+import org.apache.spark.SparkContext
+import org.apache.spark.graphx.{Edge, Graph}
+import org.apache.spark.rdd.RDD
+import thesis.Action.{CREATE, UPDATE}
+import utils.{LogUtils, UtilsUtils}
+
+import java.time.Instant
+import scala.collection.mutable
+
+object LandyConsumer {
+  def createEdge(log: LogTSV, validTo: Instant): Edge[LandyEdgePayload] = {
+    log.entity match {
+      case Entity.VERTEX(_) => throw new IllegalStateException("This should not be called on a log with vertices")
+      case Entity.EDGE(id, srcId, dstId) => {
+        val payload = LandyEdgePayload(id=id, validFrom = log.timestamp, validTo = validTo, attributes = log.attributes)
+        Edge(srcId, dstId, payload)
+      }
+    }
+  }
+
+  def createVertex(log: LogTSV, validTo: Instant): (Long, LandyVertexPayload) = {
+    log.entity match {
+      case Entity.VERTEX(objId) => {
+        val payload = LandyVertexPayload(id=objId, validFrom = log.timestamp, validTo = validTo, attributes = log.attributes)
+        (UtilsUtils.uuid, payload)
+      }
+      case Entity.EDGE(_, _, _) => throw new IllegalStateException("This should not be called on a log with edges")
+    }
+  }
+
+  def generateEdges(logs: Seq[LogTSV]): Seq[Edge[LandyEdgePayload]] = {
```

Comment:
```suggestion
def generateVertices(logs: Seq[LogTSV]): Seq[(VertexId, LandyVertexPayload)] = {
    val backwardOrderedLogs = logs.sortWith((log1, log2) => log1.timestamp.isBefore(log2.timestamp))
    val (_, vertices) = backwardOrderedLogs.foldLeft((Instant.MAX, Seq(): Seq[(VertexId, LandyVertexPayload)]))((acc, log) => {
      val (lastTimestamp, logsSoFar) = acc
      if (log.action == CREATE || log.action == UPDATE) {
        (log.timestamp, logsSoFar :+ createVertex(log, lastTimestamp))
      } else {
        (log.timestamp, logsSoFar)
      }
    })
    vertices
  }
```
Dette er ikke en suggestion til jeg vil at du skal endre. Men er et eksempel på hvordan man muligens ville gjort det med immuterbare verdier

Reply:
foldLeft er litt som reduce i javascript.

Reply:
Oooh, I like! I og med at jeg ikke er så rutta på FP og boinking synes jeg nok det blir en del komplekst, men tenker det definitivt er nice når man har litt mer erfaring på det enn meg!

Det fikk meg forsåvidt til å tenke på en podcast-episode jeg hørte på forleden som blant annet diskuterer lesbarhet mtp. nivået man er på. Lenke [her](https://podcasts.apple.com/il/podcast/episode-462-felienne-on-the-programmers-brain/id120906714?i=1000523965076), finnes også på spotify. #notsponsored

Reply:
Nice, skal sjekke ut den episoden!